banno-enrollment-po/BANNO.ENROLLMENT.V1.POW

```diff
@@ -0,0 +1,772 @@
+    DO
+     TERMINATE
+    END
+
+  END [ENDLESS LOOP]
+END [SETUP]
+
+PRINT TITLE="BANNO.ENROLLMENT.V1.POW"
+ SUPPRESSNEWLINE
+END [PRINT]
+
+PROCEDURE INITVARS
+[* Initializes variables before each run of specfile.
+*]
+ FOR PREFINDEX=0 TO PREFMAX
+  DO
+   PREFLOC(PREFINDEX)        = 0
+   PREFUSER(PREFINDEX)       = ""
+   PREFEMAILADDR(PREFINDEX)  = ""
+   PREFXFERANYLOC(PREFINDEX)     = 0  
+   PREFPASSPROB(PREFINDEX)       = 0
+  END
+ FOR ACCESSINDEX=0 TO ACCESSMAX
+  DO
+   ACCESSLOC(ACCESSINDEX)    = 0
+   ACCESSTYPE(ACCESSINDEX)   = 0
+   ACCESSACCT(ACCESSINDEX)   = ""
+   ACCESSIDTYPE(ACCESSINDEX) = 0
+   ACCESSID(ACCESSINDEX)     = ""
+   ACCESSWD(ACCESSINDEX)     = 0
+   ACCESSDEP(ACCESSINDEX)    = 0
+   ACCESSINQ(ACCESSINDEX)    = 0
+   ACCESSFM(ACCESSINDEX)     = 0
+  END
+ PREFCOUNT                   = 0
+ SELECTEDPREF                = 0
+ ACTIONREQUESTED             = 0
+ NEWPASSWORD                 = ""
+ NEWPASSPROB                 = ""
+ NEWPASSPROBNUM              = 0
+ ACCESSCOUNT                 = 0
+ SELECTEDACCESS              = 0
+ CREATEXFERTOANY             = ""
+ NEWUSERNAME                 = ""
+ NEWEMAILADDR                = ""
+ FMERROR                     = ""
+ NEWPREFLOC                  = 0
+ REALLYDELETE                = ""
+ NEWACCESSTYPE               = 0
+ NEWACCESSACCT               = ""
+ NEWACCESSIDTYPE             = 0
+ NEWACCESSID                 = ""
+ NEWACCESSWD                 = 0
+ NEWACCESSDEP                = 0
+ NEWACCESSINQ                = 0
+ NEWACCESSFM                 = 0
+ ACCESSDESC                  = ""
+END [PROCEDURE]
+
+PROCEDURE SELECTPREF
+[* Dialog that prompts user to choose banno username if more 
+** than one HB preference is present.
+*]
+ IF PREFCOUNT=1 THEN
+  SELECTEDPREF=1
+ ELSE IF PREFCOUNT>1 THEN
+  DO
+   DIALOGSTART("Select Banno User",200%,0)
+   DIALOGPROMPTLISTSTART("Choose Banno User",1)
+   FOR PREFINDEX=1 TO PREFCOUNT
+    DO
+     DIALOGPROMPTLISTOPTION(PREFINDEX,PREFUSER(PREFINDEX))
+   END
+   DIALOGPROMPTLISTEND
+   DIALOGDISPLAY
+   SELECTEDPREF=ENTERNUMBER("Choose Banno User",1)
+   DIALOGCLOSE
+  END
+END [PROCEDURE]
+
+PROCEDURE GETPREFERENCERECS
+[* Collects needed data from each HB preference on current
+** account.
+*]
+ PREFCOUNT=0
+ FOR EACH PREFERENCE WITH (PREFERENCE:HBUSERNAME<>"" AND
+                           PREFERENCE:HBMODE=STDHOMEBANKING)
+  DO
+   PREFCOUNT=PREFCOUNT+1
+   PREFLOC(PREFCOUNT)=PREFERENCE:LOCATOR
+   PREFUSER(PREFCOUNT)=PREFERENCE:HBUSERNAME
+   PREFEMAILADDR(PREFCOUNT)=PREFERENCE:EMAILADDRESS
+   PREFPASSPROB(PREFCOUNT)=PREFERENCE:PWDRESETPROBATION
+   PREFXFERANYLOC(PREFCOUNT)=0
+   FOR EACH PREFERENCE ACCESS WITH (PREFERENCE ACCESS:ACCESSTYPE=3)
+    DO
+     PREFXFERANYLOC(PREFCOUNT)=PREFERENCE ACCESS:LOCATOR
+    END
+   UNTIL PREFXFERANYLOC(PREFCOUNT)<>0
+  END
+ IF PREFCOUNT>PREFMAX THEN
+  DO
+   POPUPMESSAGE(2,"Too many users, Update Manually")
+   TERMINATE
+  END
+END [PROCEDURE]
+
+PROCEDURE CHOOSEACTION
```

Comment:
**Create/Edit Banno User options**

- I was able to create a new user without entering a password while at the same time the “Change password at next login” option was off
- If the max length of the HB Username is lowered in parameters causing some HB Usernames to be too long and the program is run to edit a user, there is no notice to the person running the program that the HB username is no longer in compliance.  If they then attempt to save the new settings, without first having shortened the HB Username to bring it in compliance, the program throws an “Unknown Prompt” exception error.

**Add/Edit Preference Access.**

- Allows for the input of an incorrect Share ID and advance to the permissions screen only to submit the permissions before being told that the ID is invalid. Seems that the program should check for the validity of the ID before displaying the Permissions screen.
- Would there be a need to have a cross account with unrestricted access while not allowing for any permissions??
- Allows for the creation of an alternate account – type share with no share ID

Reply:
1. I have added validation to ensure password is required when creating a user. I also added the same validation for user name.
2. I have added validation code that will prompt the user to change the user name if it no longer meets length criteria, then proceeds to chosen action.
3. Added code that validates account number and loan/share ID. Unfortunately, I can't interrupt the dialog flow if the account doesn't exist so the permissions screen will still come up first.
4. No, I added validation to not allow user to create access with no permissions
5. Added code that validates account number and loan/share ID.